### DISCUSSION TITLE
Confusion over understanding HOMEBREW_NO_INSTALL_UPGRADE

### DISCUSSION THREAD

**Discussion by olorton** [Homebrew/discussions#2406] [Everyday usage]
I've been trying to understand the effect of setting the HOMEBREW_NO_INSTALL_UPGRADE here: https://github.com/Homebrew/brew/blob/ed07132a3cc94110e8fbb9c412a9c9a56a51bcd4/docs/Manpage.md?plain=1#L2092

>If set, `brew install` will not automatically upgrade installed but outdated formulae

This line is confusing to me, and it feels like some words have been omitted that could clarify what changing this value actually means. For example, if I've understood this correctly, this could be re-worded to:

>If set, `brew install` will not automatically upgrade previously installed formulae, except for outdated formulae that are dependencies of a `brew install` or `brew upgrade`, will be upgraded.

Have I understood this correctly? I'm happy to submit a PR if I have.

**Comment by carlocab** [comment]
I guess that variable could be explained with some context. Suppose that you have `foo` installed, but `foo` is outdated.

Without the environment variable set, `brew install foo` will actually do `brew upgrade foo` (which includes all the consequences of doing this).

With the environment variable set, `brew install foo` does nothing.

**Comment by olorton** [comment]
Ah. I've just tested this with the following (truncated):

```
brew update
... # Updated Ruby, presumably as part of core and fetched the latest list of formulae. No upgrades were made other than Ruby.

brew outdated
... # Listed my currently outdated formulae, of which I could see Ansible.

brew install ansible
Error: ansible 4.6.0 is already installed
To upgrade to 4.8.0, run:
  brew upgrade ansible
# This would previously have auto-upgraded Ansible

brew upgrade ansible
# This upgraded Ansible and it's dependancies, but not the full list of outdated formulae.
```

As such, I propose updating the documentation to read:

>If set, `brew install` will not automatically upgrade previously installed formulae, except for outdated formulae that are dependencies of that specific `brew install`.

Unless I have miss-understood this, I shall prepare a PR to improve this.

**Comment by olorton** [comment]
I have miss-understood this, it should read:

>If set, `brew install` will not automatically upgrade previously installed formulae. Instead, use `brew upgrade formula` to upgrade a formula and it's dependencies, which will not upgrade formulae which are not dependencies. Use `brew upgrade` to upgrade all formulae.

Is that more accurate?

**Reply by carlocab** [reply]
It's an improvement, but I'm not sure it's entirely accurate. `brew upgrade foo` will not just upgrade formulae that depend on `foo`: it also upgrades formulae that `foo` depends on (both directly and indirectly), and then it will upgrade everything that depends on `foo`'s dependencies, and then also everything that those formulae depend on, and so on...

I would summarise this as "`brew upgrade foo` will upgrade everything connected to `foo` via the dependency graph", but I'm not sure that's very helpful for many people reading the manpage.

**Comment by olorton** [comment]
Another attempt:

>If set, `brew install` will not automatically upgrade previously installed formulae. Instead, use `brew upgrade formula` to upgrade a formula, and everything connected to that formula via the dependency graph. This will not upgrade formulae which are not dependencies, use `brew upgrade` to upgrade all formulae.

Furthermore:

>but I'm not sure that's very helpful for many people reading the manpage.

I disagree with your statement quoted above. It's extremely important that the manpage provide enough context to understand what the outcome of running specific commands are.

For context, I've developed a long-term frustration with the way that Homebrew has evolved to do all upgrades automatically. IMHO, installing a new package should not upgrade unrelated out of date packages on my system. This is the opposite way to how most OS package managers work. But whilst I respect the Homebrew project for having an opinion in this regard, I have been torn between taking the time to understand this project better (e.g. learning about and using flags like `HOMEBREW_NO_AUTO_UPDATE` and `HOMEBREW_NO_INSTALL_UPGRADE`) and switching to MacPorts. This recent post on [Hacker News](https://news.ycombinator.com/item?id=29079096) has encouraged me to attempt to understand ways in which Homebrew can be configured to work better for me. I think making sure the documentation makes sense is very important which is why I'm taking the time to get this wording right so that I can submit a pull request.

**Reply by carlocab** [reply]
To clarify: what I think is not helpful is describing the outcome of `brew upgrade` as

> `brew upgrade foo` will upgrade everything connected to `foo` via the dependency graph

because I think this won't be very clear to many users. I also agree with this:

> It's extremely important that the manpage provide enough context to understand what the outcome of running specific commands are.

and this:

> installing a new package should not upgrade unrelated out of date packages on my system

In fact, I think Homebrew's upgrade strategy agrees with you. I think we just have different understandings of the word "related". I think the correct definition of "related" is "connected via the dependency graph".

Unfortunately, a lot of packages are connected to each other through this, which is why it can appear that `brew upgrade foo` upgrades many unrelated packages.

In fact, I think this is inaccurate:

> This will not upgrade formulae which are not dependencies, use `brew upgrade` to upgrade all formulae.

precisely because a formula can be not-a-dependency of another formula, and yet these two formulae can be connected via the dependency graph (e.g. they might share a common dependency).

**Comment by olorton** [comment]
Thank you for your thorough response. I appreciate you highlighting how unrelated packages may have a common dependency, and so one may be upgraded as the result of upgrading another. As an aside, I think Homebrew could be improved by providing the user with a yes/no prompt to confirm which packages are upgraded (e.g. like apt-get), rather than just upgrading unrelated things automatically. I would even be happy if this was disabled by default, but could be turned on with an environment variable. However, that's not really the topic of this thread.

Where I think I disagree with you is over whether it is helpful to use terms like "dependency graph", even if it is not clear to some/many users. I didn't understand what a dependency graph was until I started using other package managers; using a package manager can cause users to look up and learn terms. I think it's far more important that the documentation be accurate and understandable; however, I take your point and whilst I agree that my proposed improvements are not quite there yet, I would argue that the current documentation

>If set, `brew install` will not automatically upgrade installed but outdated formulae

is also neither accurate nor clear. Clarifying what is going to happen with this flag enabled is my main objective here. 

Taking all of your feedback into account, and given my desire to improve this documentation, would you please review this attempt at a re-write of this:

>If set, `brew install` will not automatically upgrade previously installed formulae unless they are dependent on, or a dependency of, another formula that needs upgrading, either directly or indirectly.

**Reply by carlocab** [reply]
I agree that the explanation in the manpage should be improved. Thanks for taking the time to try to do so. However,

> If set, `brew install` will not automatically upgrade previously installed formulae unless they are dependent on, or a dependency of, another formula that needs upgrading, either directly or indirectly.

I think that this also describes the behaviour of `brew install` if the variable is not set, so I'm not sure this is a much better description either.

**Reply by olorton** [reply]
Could you (or someone else) suggest a wording for this that would better represent what is actually happening?

**Reply by MikeMcQuaid** [reply]
I've opened https://github.com/Homebrew/brew/pull/12413 to adjust the wording and the behaviour
